drivers/hmis_csv_importer/app/jobs/hmis_csv_importer/cleanup/expire_base_job.rb

```diff
@@ -0,0 +1,109 @@
+###
+# Copyright 2016 - 2024 Green River Data Analysis, LLC
+#
+# License detail: https://github.com/greenriver/hmis-warehouse/blob/production/LICENSE.md
+###
+
+# Base class for loader and importer expiration jobs
+#   Subclasses should define these methods
+#   def log_id_field
+#   end
+#
+#   def log_model
+#   end
+#
+#   def models
+#   end
+#
+module HmisCsvImporter::Cleanup
+  class ExpireBaseJob < ApplicationJob
+    include ReportingConcern
+
+    # @param data_source_id [Integer] the ID of the data source
+    # @param retain_log_count [Integer] the number of retained log records
+    # @param retain_after_date [DateTime] the date after which records are retained
+    def perform(data_source_id: nil, retain_log_count: 10, retain_after_date: DateTime.current)
+      @data_source_id = data_source_id
+      @retain_log_count = retain_log_count * files_per_log
```

Comment:
So, I'm afraid I lead us down an unnecessary path.  When we were basing the retention on the existence of data in the warehouse, we needed to know the data source and we needed to be careful about the number of files that were loaded at import time, but now that you've pivoted to looking at the loader and importer records, we can just preserve the last N (and if we want to make sure we have at least X weeks worth that's fine too.)  

We no longer need to care about the data source or number of files per import because a given row will only exist in one of those anyway, so by keeping the last N we'll always have the last N for any given row, which is sufficient, and way less complicated, than what's there now.

Let me know if that doesn't make sense, or if you think we still need this complexity.